File: common/src/main/java/dev/obscuria/fragmentum/core/v1/common/BaseModBridge.java
package dev.obscuria.fragmentum.core.v1.common;

import dev.obscuria.fragmentum.api.Fragmentum;
import dev.obscuria.fragmentum.api.v1.common.ModBridge;
import org.jetbrains.annotations.ApiStatus;

import java.util.Optional;
import java.util.function.Supplier;

@ApiStatus.Internal
public record BaseModBridge(String id, String displayName) implements ModBridge
{
    @Override
    public boolean isLoaded()
    {
        return Fragmentum.PLATFORM.isModLoaded(id);
    }

    @Override
    public <T> Optional<T> safeGet(Supplier<Supplier<T>> supplier)
    {
        return isLoaded()
                ? Optional.ofNullable(supplier.get().get())
                : Optional.empty();
    }

    @Override
    public void safeRun(Supplier<Runnable> supplier)
    {
        if (!isLoaded()) return;
        supplier.get().run();
    }

    @Override
    public void ifMissing(Runnable runnable)
    {
        if (isLoaded()) return;
        runnable.run();
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/core/client/BaseModBridge.kt
package dev.obscuria.fragmentum.core.client

import dev.obscuria.fragmentum.api.Fragmentum
import dev.obscuria.fragmentum.api.ModBridge
import java.util.function.Supplier

@JvmRecord
data class BaseModBridge(override val id: String, override val displayName: String) : ModBridge
{
    override val isLoaded: Boolean
        get() = Fragmentum.PLATFORM.isModLoaded(id)

    override fun <T> extract(supplier: Supplier<Supplier<T>>): T?
    {
        return if (isLoaded) return supplier.get().get() else null
    }

    override fun invoke(supplier: Supplier<Runnable>): ModBridge
    {
        if (!isLoaded) return this
        supplier.get().run()
        return this
    }

    override fun fallback(runnable: Runnable): ModBridge
    {
        if (isLoaded) return this
        runnable.run()
        return this
    }
}